File: packages/backend/transport/src/transport-repository.ts
/**
 * Transport Repository Interface
 *
 * Defines the data access contract for transport operations.
 * Implementations can use Prisma, in-memory stores, or other backends.
 *
 * Requirements: 1.2 (Transport_Module)
 */
import type { PaginationOptions, PaginatedResult } from '@proctira/common';

// ─── Transport Route Entity ──────────────────────────────────────────────────

export type RouteStatus = 'active' | 'inactive' | 'suspended';

export interface TransportRouteEntity {
  id: string;
  tenantId: string;
  name: string;
  description: string | null;
  status: RouteStatus;
  startLocation: string;
  endLocation: string;
  distanceKm: number | null;
  estimatedDurationMinutes: number | null;
  operatingDays: string[];
  departureTime: string | null;
  returnTime: string | null;
  institutionId: string | null;
  createdAt: Date;
  updatedAt: Date;
}

export interface RouteFilter {
  status?: RouteStatus;
  institutionId?: string;
  search?: string;
}

// ─── Route Stop Entity ───────────────────────────────────────────────────────

export interface RouteStopEntity {
  id: string;
  tenantId: string;
  routeId: string;
  name: string;
  latitude: number | null;
  longitude: number | null;
  stopOrder: number;
  pickupTime: string | null;
  dropoffTime: string | null;
  createdAt: Date;
  updatedAt: Date;
}

// ─── Vehicle Entity ──────────────────────────────────────────────────────────

export type VehicleStatus = 'active' | 'inactive' | 'maintenance' | 'retired';

export interface VehicleEntity {
  id: string;
  tenantId: string;
  registrationNumber: string;
  make: string | null;
  model: string | null;
  year: number | null;
  capacity: number;
  status: VehicleStatus;
  insuranceExpiry: string | null;
  lastServiceDate: string | null;
  createdAt: Date;
  updatedAt: Date;
}

export interface VehicleFilter {
  status?: VehicleStatus;
  search?: string;
}

// ─── Driver Assignment Entity ────────────────────────────────────────────────

export interface DriverAssignmentEntity {
  id: string;
  tenantId: string;
  vehicleId: string;
  driverId: string;
  routeId: string | null;
  startDate: string;
  endDate: string | null;
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;
}

export interface DriverAssignmentFilter {
  vehicleId?: string;
  driverId?: string;
  routeId?: string;
  isActive?: boolean;
}

// ─── Student Assignment Entity ───────────────────────────────────────────────

export interface StudentRouteAssignmentEntity {
  id: string;
  tenantId: string;
  studentId: string;
  routeId: string;
  stopId: string | null;
  startDate: string;
  endDate: string | null;
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;
}

export interface StudentAssignmentFilter {
  studentId?: string;
  routeId?: string;
  stopId?: string;
  isActive?: boolean;
}

// ─── Wave 9 / G-920 ops entities ──────────────────────────────────────────────

export interface VehicleDeviceEntity {
  id: string;
  tenantId: string;
  vehicleId: string;
  deviceId: string;
  deviceKeyHash: string;
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;
}

export interface GpsPingEntity {
  id: string;
  tenantId: string;
  vehicleId: string;
  deviceId: string;
  pingId: string;
  latitude: number;
  longitude: number;
  recordedAt: Date;
  speedKph: number | null;
  headingDeg: number | null;
  createdAt: Date;
}

export type TripDirection = 'pickup' | 'drop';
export type BusAttendanceStatus = 'boarded' | 'alighted' | 'absent';

export interface BusAttendanceEntity {
  id: string;
  tenantId: string;
  routeId: string;
  tripDate: string;
  direction: TripDirection;
  studentId: string;
  stopId: string | null;
  status: BusAttendanceStatus;
  recordedAt: Date;
  recordedBy: string | null;
  createdAt: Date;
  updatedAt: Date;
}

export type AlertKind = 'delay_minutes' | 'geofence_exit' | 'missed_pickup';

export interface AlertRuleEntity {
  id: string;
  tenantId: string;
  kind: AlertKind;
  threshold: number;
  channels: string[];
  routeId: string | null;
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;
}

export interface TransportAlertEntity {
  id: string;
  tenantId: string;
  ruleId: string;
  kind: AlertKind;
  vehicleId: string | null;
  routeId: string | null;
  studentId: string | null;
  message: string;
  payload: Record<string, unknown>;
  acknowledgedAt: Date | null;
  acknowledgedBy: string | null;
  createdAt: Date;
}

export interface TransportFeeStructureEntity {
  id: string;
  tenantId: string;
  name: string;
  routeId: string | null;
  stopId: string | null;
  minDistanceKm: number | null;
  maxDistanceKm: number | null;
  amountCents: number;
  currency: string;
  feesStructureId: string | null;
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;
}

export type TransportFeeLinkStatus = 'invoiced' | 'pending' | 'skipped';

export interface TransportFeeLinkEntity {
  id: string;
  tenantId: string;
  assignmentId: string;
  studentId: string;
  transportFeeStructureId: string | null;
  feesInvoiceId: string | null;
  feesStructureId: string | null;
  status: TransportFeeLinkStatus;
  reason: string | null;
  createdAt: Date;
}

// ─── Repository Interface ────────────────────────────────────────────────────

export interface TransportRepository {
  // Route operations
  createRoute(
    data: Omit<TransportRouteEntity, 'createdAt' | 'updatedAt'>,
  ): Promise<TransportRouteEntity>;
  updateRoute(
    id: string,
    tenantId: string,
    data: Partial<TransportRouteEntity>,
  ): Promise<TransportRouteEntity | null>;
  findRouteById(id: string, tenantId: string): Promise<TransportRouteEntity | null>;
  listRoutes(
    tenantId: string,
    filter: RouteFilter,
    pagination: PaginationOptions,
  ): Promise<PaginatedResult<TransportRouteEntity>>;
  deleteRoute(id: string, tenantId: string): Promise<boolean>;

  // Route stop operations
  createStop(data: Omit<RouteStopEntity, 'createdAt' | 'updatedAt'>): Promise<RouteStopEntity>;
  updateStop(
    id: string,
    tenantId: string,
    data: Partial<RouteStopEntity>,
  ): Promise<RouteStopEntity | null>;
  findStopById(id: string, tenantId: string): Promise<RouteStopEntity | null>;
  listStopsByRoute(routeId: string, tenantId: string): Promise<RouteStopEntity[]>;
  deleteStop(id: string, tenantId: string): Promise<boolean>;

  // Vehicle operations
  createVehicle(data: Omit<VehicleEntity, 'createdAt' | 'updatedAt'>): Promise<VehicleEntity>;
  updateVehicle(
    id: string,
    tenantId: string,
    data: Partial<VehicleEntity>,
  ): Promise<VehicleEntity | null>;
  findVehicleById(id: string, tenantId: string): Promise<VehicleEntity | null>;
  findVehicleByRegistration(
    registrationNumber: string,
    tenantId: string,
  ): Promise<VehicleEntity | null>;
  listVehicles(
    tenantId: string,
    filter: VehicleFilter,
    pagination: PaginationOptions,
  ): Promise<PaginatedResult<VehicleEntity>>;
  deleteVehicle(id: string, tenantId: string): Promise<boolean>;

  // Driver assignment operations
  createDriverAssignment(
    data: Omit<DriverAssignmentEntity, 'createdAt' | 'updatedAt'>,
  ): Promise<DriverAssignmentEntity>;
  updateDriverAssignment(
    id: string,
    tenantId: string,
    data: Partial<DriverAssignmentEntity>,
  ): Promise<DriverAssignmentEntity | null>;
  findDriverAssignmentById(id: string, tenantId: string): Promise<DriverAssignmentEntity | null>;
  listDriverAssignments(
    tenantId: string,
    filter: DriverAssignmentFilter,
    pagination: PaginationOptions,
  ): Promise<PaginatedResult<DriverAssignmentEntity>>;
  findActiveDriverAssignment(
    vehicleId: string,
    tenantId: string,
  ): Promise<DriverAssignmentEntity | null>;

  // Student route assignment operations
  createStudentAssignment(
    data: Omit<StudentRouteAssignmentEntity, 'createdAt' | 'updatedAt'>,
  ): Promise<StudentRouteAssignmentEntity>;
  updateStudentAssignment(
    id: string,
    tenantId: string,
    data: Partial<StudentRouteAssignmentEntity>,
  ): Promise<StudentRouteAssignmentEntity | null>;
  findStudentAssignmentById(
    id: string,
    tenantId: string,
  ): Promise<StudentRouteAssignmentEntity | null>;
  listStudentAssignments(
    tenantId: string,
    filter: StudentAssignmentFilter,
    pagination: PaginationOptions,
  ): Promise<PaginatedResult<StudentRouteAssignmentEntity>>;
  findActiveStudentAssignment(
    studentId: string,
    tenantId: string,
  ): Promise<StudentRouteAssignmentEntity | null>;

  // Wave 9 / G-920 ops
  listAllStops(tenantId: string): Promise<RouteStopEntity[]>;

  registerVehicleDevice(
    data: Omit<VehicleDeviceEntity, 'createdAt' | 'updatedAt'>,
  ): Promise<VehicleDeviceEntity>;
  findDeviceByDeviceId(deviceId: string, tenantId: string): Promise<VehicleDeviceEntity | null>;
  findDeviceByVehicleId(vehicleId: string, tenantId: string): Promise<VehicleDeviceEntity | null>;

  ingestGpsPing(
    data: Omit<GpsPingEntity, 'createdAt'>,
  ): Promise<{ ping: GpsPingEntity; duplicate: boolean }>;
  listGpsPingsForVehicle(tenantId: string, vehicleId: string): Promise<GpsPingEntity[]>;
  listLatestGpsPingPerVehicle(tenantId: string): Promise<GpsPingEntity[]>;

  upsertBusAttendance(
    data: Omit<BusAttendanceEntity, 'createdAt' | 'updatedAt'>,
  ): Promise<BusAttendanceEntity>;
  listBusAttendanceTrip(
    tenantId: string,
    filter: { routeId: string; tripDate: string; direction: TripDirection },
  ): Promise<BusAttendanceEntity[]>;

  createAlertRule(data: Omit<AlertRuleEntity, 'createdAt' | 'updatedAt'>): Promise<AlertRuleEntity>;
  listAlertRules(tenantId: string): Promise<AlertRuleEntity[]>;
  createAlert(data: Omit<TransportAlertEntity, 'createdAt'>): Promise<TransportAlertEntity>;
  listAlerts(tenantId: string): Promise<TransportAlertEntity[]>;
  acknowledgeAlert(
    id: string,
    tenantId: string,
    acknowledgedBy: string,
  ): Promise<TransportAlertEntity | null>;

  createTransportFeeStructure(
    data: Omit<TransportFeeStructureEntity, 'createdAt' | 'updatedAt'>,
  ): Promise<TransportFeeStructureEntity>;
  listTransportFeeStructures(tenantId: string): Promise<TransportFeeStructureEntity[]>;
  createFeeLink(data: Omit<TransportFeeLinkEntity, 'createdAt'>): Promise<TransportFeeLinkEntity>;
  listFeeLinks(tenantId: string): Promise<TransportFeeLinkEntity[]>;
  findFeeLinkByAssignment(
    assignmentId: string,
    tenantId: string,
  ): Promise<TransportFeeLinkEntity | null>;
}

File: apps/mobile/lib/features/transport/data/transport_repository.dart
import 'package:dio/dio.dart';

class TransportRoute {
  const TransportRoute({
    required this.id,
    required this.name,
    required this.status,
    required this.startLocation,
    required this.endLocation,
    this.description,
    this.distanceKm,
    this.estimatedDurationMinutes,
    this.departureTime,
    this.returnTime,
  });

  final String id;
  final String name;
  final String status;
  final String startLocation;
  final String endLocation;
  final String? description;
  final num? distanceKm;
  final int? estimatedDurationMinutes;
  final String? departureTime;
  final String? returnTime;

  factory TransportRoute.fromJson(Map<String, dynamic> json) {
    return TransportRoute(
      id: json['id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      status: json['status'] as String? ?? 'active',
      startLocation: json['startLocation'] as String? ?? '',
      endLocation: json['endLocation'] as String? ?? '',
      description: json['description'] as String?,
      distanceKm: json['distanceKm'] as num?,
      estimatedDurationMinutes: json['estimatedDurationMinutes'] as int?,
      departureTime: json['departureTime'] as String?,
      returnTime: json['returnTime'] as String?,
    );
  }
}

class TransportStop {
  const TransportStop({
    required this.id,
    required this.name,
    required this.stopOrder,
    this.pickupTime,
    this.dropoffTime,
  });

  final String id;
  final String name;
  final int stopOrder;
  final String? pickupTime;
  final String? dropoffTime;

  factory TransportStop.fromJson(Map<String, dynamic> json) {
    return TransportStop(
      id: json['id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      stopOrder: json['stopOrder'] as int? ?? 0,
      pickupTime: json['pickupTime'] as String?,
      dropoffTime: json['dropoffTime'] as String?,
    );
  }
}

class TransportVehicle {
  const TransportVehicle({
    required this.id,
    required this.registrationNumber,
    required this.capacity,
    required this.status,
    this.make,
    this.model,
    this.insuranceExpiry,
  });

  final String id;
  final String registrationNumber;
  final int capacity;
  final String status;
  final String? make;
  final String? model;
  final String? insuranceExpiry;

  factory TransportVehicle.fromJson(Map<String, dynamic> json) {
    return TransportVehicle(
      id: json['id'] as String? ?? '',
      registrationNumber: json['registrationNumber'] as String? ?? '',
      capacity: json['capacity'] as int? ?? 0,
      status: json['status'] as String? ?? 'active',
      make: json['make'] as String?,
      model: json['model'] as String?,
      insuranceExpiry: json['insuranceExpiry'] as String?,
    );
  }
}

class StudentRouteAssignment {
  const StudentRouteAssignment({
    required this.id,
    required this.studentId,
    required this.routeId,
    required this.startDate,
    required this.isActive,
    this.stopId,
    this.endDate,
  });

  final String id;
  final String studentId;
  final String routeId;
  final String startDate;
  final bool isActive;
  final String? stopId;
  final String? endDate;

  factory StudentRouteAssignment.fromJson(Map<String, dynamic> json) {
    return StudentRouteAssignment(
      id: json['id'] as String? ?? '',
      studentId: json['studentId'] as String? ?? '',
      routeId: json['routeId'] as String? ?? '',
      startDate: json['startDate'] as String? ?? '',
      isActive: json['isActive'] as bool? ?? true,
      stopId: json['stopId'] as String?,
      endDate: json['endDate'] as String?,
    );
  }
}

class TransportOverview {
  const TransportOverview({
    required this.routes,
    required this.vehicles,
    required this.assignments,
  });

  final List<TransportRoute> routes;
  final List<TransportVehicle> vehicles;
  final List<StudentRouteAssignment> assignments;

  int get activeRouteCount =>
      routes.where((TransportRoute r) => r.status == 'active').length;

  int get activeVehicleCount =>
      vehicles.where((TransportVehicle v) => v.status == 'active').length;

  int get activeAssignmentCount => assignments
      .where((StudentRouteAssignment a) => a.isActive)
      .length;
}

/// Dio client for `/api/v1/transport/*`.
class TransportRepository {
  TransportRepository({required Dio dio}) : _dio = dio;

  final Dio _dio;

  Future<TransportOverview> getOverview() async {
    final List<TransportRoute> routes = await listRoutes();
    final List<TransportVehicle> vehicles = await listVehicles();
    final List<StudentRouteAssignment> assignments = await listAssignments();
    return TransportOverview(
      routes: routes,
      vehicles: vehicles,
      assignments: assignments,
    );
  }

  Future<List<TransportRoute>> listRoutes({String? search}) async {
    try {
      final Response<dynamic> response = await _dio.get(
        '/api/v1/transport/routes',
        queryParameters: <String, dynamic>{
          'pageSize': 100,
          if (search != null && search.isNotEmpty) 'search': search,
        },
      );
      return _unwrapList(response.data)
          .map(TransportRoute.fromJson)
          .toList(growable: false);
    } on DioException {
      return const <TransportRoute>[];
    }
  }

  Future<TransportRoute?> getRoute(String id) async {
    try {
      final Response<dynamic> response =
          await _dio.get('/api/v1/transport/routes/$id');
      final Object? raw = response.data;
      if (raw is! Map) return null;
      return TransportRoute.fromJson(Map<String, dynamic>.from(raw));
    } on DioException {
      return null;
    }
  }

  Future<List<TransportStop>> listStops(String routeId) async {
    try {
      final Response<dynamic> response =
          await _dio.get('/api/v1/transport/routes/$routeId/stops');
      return _unwrapList(response.data)
          .map(TransportStop.fromJson)
          .toList(growable: false);
    } on DioException {
      return const <TransportStop>[];
    }
  }

  Future<List<TransportVehicle>> listVehicles() async {
    try {
      final Response<dynamic> response = await _dio.get(
        '/api/v1/transport/vehicles',
        queryParameters: const <String, dynamic>{'pageSize': 100},
      );
      return _unwrapList(response.data)
          .map(TransportVehicle.fromJson)
          .toList(growable: false);
    } on DioException {
      return const <TransportVehicle>[];
    }
  }

  Future<List<StudentRouteAssignment>> listAssignments({String? routeId}) async {
    try {
      final Response<dynamic> response = await _dio.get(
        '/api/v1/transport/student-assignments',
        queryParameters: <String, dynamic>{
          'pageSize': 100,
          'routeId': ?routeId,
        },
      );
      return _unwrapList(response.data)
          .map(StudentRouteAssignment.fromJson)
          .toList(growable: false);
    } on DioException {
      return const <StudentRouteAssignment>[];
    }
  }

  List<Map<String, dynamic>> _unwrapList(dynamic payload) {
    if (payload == null) return const <Map<String, dynamic>>[];
    if (payload is List) {
      return payload
          .whereType<Map>()
          .map((Map e) => Map<String, dynamic>.from(e))
          .toList(growable: false);
    }
    if (payload is Map && payload['data'] is List) {
      return (payload['data'] as List)
          .whereType<Map>()
          .map((Map e) => Map<String, dynamic>.from(e))
          .toList(growable: false);
    }
    return const <Map<String, dynamic>>[];
  }
}
